### ISSUE TITLE
Support KEP 753 Kubernetes Sidecar Containers

### ISSUE THREAD

**Issue by danehans** [istio/istio#44537]
**Describe the feature request**
https://github.com/kubernetes/enhancements/issues/753 tracks the need for K8s to provide improved support for the sidecar container pattern. The associated [KEP](https://github.com/kubernetes/enhancements/pull/3761) has merged and feature implementation is underway. Istio should consider supporting the [SidecarContainers feature](https://github.com/kubernetes/kubernetes/pull/116429). 

**Describe alternatives you've considered**
Continue as-is and not support the feature.

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[X] Docs
[X] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**
https://github.com/istio/istio/issues/6324 is a similar similar feature but is closed and specific to batch jobs.

**Comment by keithmattix** [comment]
The feature will still be alpha when the implementation merged (presumably in 1.28) correct?

**Comment by danehans** [comment]
Yes, I anticipate the feature being in k8s 1.28. I'm unsure of the Istio policy for supporting k8s features. Does a k8s feature need to be beta or stable for it to be approved for implementation? I recall Gateway API being implemented when the APIs were alpha.

**Comment by keithmattix** [comment]
Slight difference between Gateway API and the sidecar KEP. Gateway API is installed via CRDs while the sidecar KEP modifies core k8s resources (e.g. Pod). That means that the functionality will be behind a feature gate that is off by default. AKS doesn't currently have a way to enable alpha k8s features, and in GKE, you have to create a separate, alpha only cluster to do it. So it's not likely that there will be a ton of usage for this

**Comment by howardjohn** [comment]
All tests passing in https://github.com/istio/istio/pull/44835

**Comment by keithmattix** [comment]
Hidden behind feature flag in #45959; leaving this open until the KEP gets past alpha in k8s

**Comment by martynd** [comment]
In 1.29 this goes beta and is enabled by default. GKE is scheduled to release 1.29 on the rapid channel this Friday

**Comment by howardjohn** [comment]
I should probably write a proper doc, but for now you need to set `ENABLE_NATIVE_SIDECARS` env var in istiod to use this btw. Note its your responsibility to ensure all _nodes_ are the appropriate version (1.29), not just the conrtol plane.

If you try it out, let us know how it goes!

**Comment by avisaradir** [comment]
> I should probably write a proper doc, but for now you need to set `ENABLE_NATIVE_SIDECARS` env var in istiod to use this btw. Note its your responsibility to ensure all _nodes_ are the appropriate version (1.29), not just the conrtol plane.
> 
> If you try it out, let us know how it goes!

Using helm how do I enable `ENABLE_NATIVE_SIDECARS` env this variable?

**Comment by hanxiaop** [comment]
> > I should probably write a proper doc, but for now you need to set `ENABLE_NATIVE_SIDECARS` env var in istiod to use this btw. Note its your responsibility to ensure all _nodes_ are the appropriate version (1.29), not just the conrtol plane.
> > If you try it out, let us know how it goes!
> 
> Using helm how do I enable `ENABLE_NATIVE_SIDECARS` env this variable?

@avisaradir  `--set pilot.env.ENABLE_NATIVE_SIDECARS=true`, and make sure your k8s nodes support it. See https://istio.io/latest/blog/2023/native-sidecars/

**Comment by avisaradir** [comment]
> > > I should probably write a proper doc, but for now you need to set `ENABLE_NATIVE_SIDECARS` env var in istiod to use this btw. Note its your responsibility to ensure all _nodes_ are the appropriate version (1.29), not just the conrtol plane.
> > > If you try it out, let us know how it goes!
> > 
> > 
> > Using helm how do I enable `ENABLE_NATIVE_SIDECARS` env this variable?
> 
> @avisaradir `--set pilot.env.ENABLE_NATIVE_SIDECARS=true`, and make sure your k8s nodes support it.

Thank you, figured it mid night. 
Works prefect disabled my mutation webhook to force the quit command.

**Comment by martynd** [comment]
> I should probably write a proper doc, but for now you need to set `ENABLE_NATIVE_SIDECARS` env var in istiod to use this btw. Note its your responsibility to ensure all _nodes_ are the appropriate version (1.29), not just the conrtol plane.
> 
> If you try it out, let us know how it goes!

Ended up having to wait until it hit the regular channel to get it on our cluster (and another week for it to reach us-central1 😮‍💨) + the inevitable node pool cycles, but we got there in the end!
Haven't experienced any issues so far (on Istio's side at least 👍).
 
As a possible caution to others:
We are having troubles with the GKE autoscaler (and have a ticket open)
It seems its not properly aware of the feature gate despite all the pools being on 1.29. The autoscaler detects the need for the SidecarContainers gate but then refuses to scale anything up because it thinks its disabled.
Manually scaling up the pools then allows the pods to schedule, but obviously the autoscaler only being able to scale down poses issues.

In the coming weeks I'm hoping to give the CNI driver a go in tandem with the new sidecars
